src/Workspaces/Core/Portable/Formatting/Engine/AbstractFormatEngine.cs

```diff
@@ -271,36 +271,41 @@ void beginningOfTreeTriviaInfoApplier(int i, TriviaData info)
                 return;
             }
 
-            void endOfTreeTriviaInfoApplier(int i, TriviaData info)
+            void endOfTreeTriviaInfoApplier(int i, TokenStream ts, TriviaData info)
             {
-                tokenStream.ApplyEndOfTreeChange(info);
+                ts.ApplyEndOfTreeChange(info);
             }
 
             // remove all trailing indentation
             var triviaInfo = tokenStream.GetTriviaDataAtEndOfTree().WithIndentation(0, context, _formattingRules, cancellationToken);
 
-            triviaInfo.Format(context, _formattingRules, endOfTreeTriviaInfoApplier, cancellationToken);
+            triviaInfo.Format(context, _formattingRules, endOfTreeTriviaInfoApplier, tokenStream, cancellationToken);
         }
 
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/30819", AllowCaptures = false)]
         private void ApplyTriviaOperations(FormattingContext context, TokenStream tokenStream, CancellationToken cancellationToken)
         {
-            // trivia formatting result appliers
-            void regularApplier(int tokenPairIndex, TriviaData info)
+            void regularApplier(int tokenPairIndex, TriviaData info, TokenStream ts)
             {
-                tokenStream.ApplyChange(tokenPairIndex, info);
+                ts.ApplyChange(tokenPairIndex, info);
             }
 
-            // trivia formatting applier
-            void triviaFormatter(int tokenPairIndex)
+            void triviaFormatter(int tokenPairIndex, FormattingContext ctx, ChainedFormattingRules formattingRules, TokenStream ts, CancellationToken ct)
             {
-                var triviaInfo = tokenStream.GetTriviaData(tokenPairIndex);
-                triviaInfo.Format(context, _formattingRules, regularApplier, cancellationToken, tokenPairIndex);
+                var triviaInfo = ts.GetTriviaData(tokenPairIndex);
+                triviaInfo.Format(
+                    ctx,
+                    formattingRules,
+                    (int tokenPairIndex1, TokenStream ts2, TriviaData info) => regularApplier(tokenPairIndex1, info, ts2),
```

Comment:
lambda probably isn't needed?

Reply:
As far as I can tell, the delegate isn't cached even for local functions unless the lambda is used.